kubernetes/annotations_and_labels.md

```diff
@@ -160,6 +182,156 @@ This page defines common annotations and labels we set in Kubernetes objects.
 
 #### Tenant Cluster
 
+### Labels Set On Objects Installed By Charts
+
+Labels related to a higher-level virtual concept of an _app_, i.e. a bunch of
+components usually installed by one or more Helm charts.
+
+- `app`
+- `app.kubernetes.io/name`
+- `app.kubernetes.io/instance`
+- `app.giantswarm.io/branch`
+- `app.giantswarm.io/commit`
+- `app.kubernetes.io/managed-by`
+- `app.kubernetes.io/version`
+- `helm.sh/chart`
+
+To set those labels without having to repeat their definition in multiple
+places we use a template helper based on that in [chart template][helm-def-tpl]
+from upstream Helm. This template helper is in a form of a file `_helpers.tpl`
+that should be placed inside the chart's `templates/` subdirectory and it
+defines a couple of variables that can then be re-used throughout other
+templates in the chart.
+
+Contents of `_helpers.tpl` (replace `aws-operator` with the name of your
+app/operator):
+
+``` HTML+Django
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "aws-operator.name" -}}
+{{- .Chart.Name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "aws-operator.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Common labels
+*/}}
+{{- define "aws-operator.labels" -}}
+app: {{ include "aws-operator.name" . | quote }}
+{{ include "aws-operator.selectorLabels" . }}
+app.giantswarm.io/branch: {{ .Values.project.branch | quote }}
+app.giantswarm.io/commit: {{ .Values.project.commit | quote }}
+app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
+helm.sh/chart: {{ include "aws-operator.chart" . | quote }}
+{{- end -}}
+
+{{/*
+Selector labels
+*/}}
+{{- define "aws-operator.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "aws-operator.name" . | quote }}
+app.kubernetes.io/instance: {{ .Release.Name | quote }}
+{{- end -}}
+```
+
+This defines a couple of strings based on chart name/version normalised to
+comply with k8s resource name [constraints][k8s-identifiers], and two snippets
+with labels:
+
+- `aws-operator.name` - name of the chart, trimmed to 63 characters
+- `aws-operator.chart` - normalised name + version of the chart, i.e. trimmed
+  to 63 characters and with `+` signs replaced with `-`
+- `aws-operator.labels` - defines all the labels described above, including the
+  selector labels; usage: `{{- include "aws-operator.labels" . | nindent 4 }}`
+  (adjust indent as required)
```

Comment:
To avoid copy/paste errors.

```suggestion
  selector labels; usage: `{{- include "aws-operator.labels" . | nindent INDENT }}`
  (set INDENT to required integer)
```

Reply:
Done